### PR TITLE
[Docs] Add runner tutorial

### PR DESCRIPTION
**Pull request by LXXXXR**
## Motivation
Add runner tutorial.

## Modification
Add `docs/runner.md`

### CODE REVIEW

**Comment by LXXXXR** [comment]
@zhouzaida @ZwwWayne 
Pls kindly take a look. Thank you.

File: docs/runner.md [hunk 1]
```diff
@@ -1,6 +1,158 @@
 ## Runner
 
-The runner module aims to help users to start training with less code, while stays
-flexible and configurable.
+The runner class is designed to manage the training. It eases the training process with less code demanded from users while staying flexible and configurable. The main features are as listed:
 
-Documentation and examples are still on going.
+- Support `EpochBasedRunner` and `IterBasedRunner` for different scenarios.
```
Comment (ZwwWayne) [inline docs/runner.md:5]: The users are also able to implement new runners to meet their own needs.

File: docs/runner.md [hunk 2]
```diff
@@ -1,6 +1,158 @@
...
+- Support customized workflow to allow switching between different modes while training. Currently, supported modes are train and val.
+- Enable extensibility through various Hooks, including hooks defined in MMCV and customized ones.
+
+## EpochBasedRunner
+
+As its name indicates, workflow in `EpochBasedRunner` should be set based on epochs. For example, [('train', 2), ('val', 1)] means running 2 epochs for training and 1 epoch for validation, iteratively. And each epoch may contain multiple iters. Currently, MMDetection uses `EpochBasedRunner` by default.
```
Comment (ZwwWayne) [inline docs/runner.md:11]: iters -> iteration.

File: docs/runner.md [hunk 3]
```diff
@@ -1,6 +1,158 @@
...
+Currently, we support 2 modes: train and val. Let's take a look at their core logic:
+
+```python
+# Currently, epoch_runner could be either train or val
+def epoch_runner(self, data_loader, **kwargs):
+    # traverse the dataset and get batch data for 1 epoch
+    for i, data_batch in enumerate(data_loader):
+        self.call_hook('before_train(val)_iter')
+        self.run_iter(data_batch, train_mode=True or False, **kwargs)
+        self.call_hook('after_train(val)_iter')
```
Comment (ZwwWayne) [inline docs/runner.md:38]: This part of code is a little bit vague `train(val)`. Use pseudo-code if necessary.

Comment (LXXXXR) [reply docs/runner.md]: Does it look clearer now?

File: docs/runner.md [hunk 4]
```diff
@@ -1,6 +1,159 @@
 ## Runner
 
-The runner module aims to help users to start training with less code, while stays
-flexible and configurable.
+The runner class is designed to manage the training. It eases the training process with less code demanded from users while staying flexible and configurable. The main features are as listed:
 
-Documentation and examples are still on going.
+- Support `EpochBasedRunner` and `IterBasedRunner` for different scenarios. Implementing customized runners is also allowed to meet customized needs.
+- Support customized workflow to allow switching between different modes while training. Currently, supported modes are train and val.
+- Enable extensibility through various Hooks, including hooks defined in MMCV and customized ones.
```
Comment (ZwwWayne) [inline docs/runner.md:7]: Hooks or hooks?

File: docs/runner.md [hunk 5]
```diff
@@ -1,6 +1,159 @@
...
+    for i, flow in enumerate(workflow):
+        # mode(e.g. train) determines which function to run
+        mode, epochs = flow
+        epoch_runner = getattr(self, mode)
+        # execute the corresponding function
+        for _ in range(epochs):
+            epoch_runner(data_loaders[i], **kwargs)
+```
+
+Currently, we support 2 modes: train and val. Let's take train function for example and have a look at its core logic:
```
Comment (ZwwWayne) [inline docs/runner.md:29]: a train function

File: docs/runner.md [hunk 6]
```diff
@@ -1,6 +1,159 @@
...
+
+```python
+# the condition to stop training
+while curr_epoch < max_epochs:
+    # traverse the workflow.
+    # e.g. workflow = [('train', 2), ('val', 1)]
+    for i, flow in enumerate(workflow):
+        # mode(e.g. train) determines which function to run
+        mode, epochs = flow
+        epoch_runner = getattr(self, mode)
```
Comment (ZwwWayne) [inline docs/runner.md:23]: may explain what epoch_runner will be, e.g., it will be self.train() or self.val()

Comment (ZwwWayne) [reply docs/runner.md:23]: The part of self.call_hook can also be explained here.

File: docs/runner.md [hunk 7]
```diff
@@ -1,6 +1,159 @@
...
+
+```python
+# Currently, epoch_runner could be either train or val
+def train(self, data_loader, **kwargs):
+    # traverse the dataset and get batch data for 1 epoch
+    for i, data_batch in enumerate(data_loader):
+        self.call_hook('before_train_iter')
+        # set train_mode as False in val function
+        self.run_iter(data_batch, train_mode=True, **kwargs)
+        self.call_hook('after_train_iter')
```
Comment (ZwwWayne) [inline docs/runner.md:39]: Explain what will self.call_hook will do, e.g., it will pass all its hooks and execute their `after_train_iter` function.

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**